src/gui/qgsmapcanvas.cpp

```diff
@@ -604,8 +604,10 @@ void QgsMapCanvas::stopRendering()
   {
     QgsDebugMsg( "CANVAS stop rendering!" );
     mJobCanceled = true;
-    mJob->cancel();
-    Q_ASSERT( !mJob ); // no need to delete here: already deleted in finished()
+    disconnect( mJob, &QgsMapRendererJob::finished, this, &QgsMapCanvas::rendererJobFinished );
+    connect( mJob, &QgsMapRendererQImageJob::finished, mJob, &QgsMapRendererQImageJob::deleteLater );
+    mJob->cancelWithoutBlocking();
+    mJob = nullptr;
```

Comment:
If the application exits, threads running in the background can easily trigger crashes, regardless of connected signals I think. Or did I get something wrong there?

Cached images, possibly yes. The tests failed on me, so if it was still a problem now you'd have noticed :)